### PR TITLE
EVG-5948 remove task cache

### PR DESCRIPTION
**Pull request by ybrill**
[EVG-5948](https://jira.mongodb.org/browse/EVG-5948)

### Description 
Most of the build's task cache has been unreferenced since #4009. This PR is to stop making updates to it.
The task _id_ slice is still useful, so I'm leaving that. It's created when the build is created and only updated by generate tasks.

### CODE REVIEW

File: model/lifecycle.go [hunk 1]
```diff
@@ -248,14 +224,10 @@ func SetBuildPriority(buildId string, priority int64, caller string) error {
 		if err != nil {
 			return errors.Wrapf(err, "can't get tasks for build '%s'", buildId)
 		}
-		var deactivatedTasks []task.Task
-		deactivatedTasks, err = task.DeactivateTasks(tasks, caller)
+		err = task.DeactivateTasks(tasks, caller)
```
Comment (ablack12) [inline model/lifecycle.go:227]: nit: now that this has only an error to return can we combine with with the if statement

File: model/lifecycle_test.go [hunk 2]
```diff
@@ -1038,71 +934,6 @@ func TestCreateBuildFromVersion(t *testing.T) {
 
 		})
 
-		Convey("the build should contain task caches that correspond exactly"+
```
Comment (ablack12) [inline model/lifecycle_test.go:1041]: This test still seems useful, but just checking only tasks[i].Id

File: model/lifecycle_test.go [hunk 3]
```diff
@@ -1038,71 +934,6 @@ func TestCreateBuildFromVersion(t *testing.T) {
...
-			So(build.Tasks[1].Status, ShouldEqual, evergreen.TaskUndispatched)
-			So(build.Tasks[2].Id, ShouldNotEqual, "")
-			So(build.Tasks[2].DisplayName, ShouldEqual, "taskC")
-			So(build.Tasks[2].Status, ShouldEqual, evergreen.TaskUndispatched)
-			So(build.Tasks[3].Id, ShouldNotEqual, "")
-			So(build.Tasks[3].DisplayName, ShouldEqual, "taskE")
-			So(build.Tasks[3].Status, ShouldEqual, evergreen.TaskUndispatched)
-		})
-
-		Convey("a task cache should not contain execution tasks that are part of a display task", func() {
```
Comment (ablack12) [inline model/lifecycle_test.go:1074]: assuming this condition is still true, we should also keep this test (although probably checking ID specifically now)

File: model/task_lifecycle.go [hunk 4]
```diff
@@ -62,7 +60,7 @@ func SetActiveState(t *task.Task, caller string, active bool) error {
 			tasksToActivate = append(tasksToActivate, originalTasks...)
 		}
 		var err error
-		modifiedTasks, err = task.ActivateTasks(tasksToActivate, time.Now(), caller)
+		err = task.ActivateTasks(tasksToActivate, time.Now(), caller)
```
Comment (ablack12) [inline model/task_lifecycle.go:63]: same comment here about one lining

**Review by ablack12 — CHANGES_REQUESTED** [review]
[This](https://github.com/evergreen-ci/evergreen/blob/2a5e1e89e8d581d1a841b216b7ad1209f20cb586/public/static/js/build.js#L226) and [this](https://github.com/evergreen-ci/evergreen/blob/2a5e1e89e8d581d1a841b216b7ad1209f20cb586/public/static/js/build.js#L25) seem potentially related to the task cache

File: model/lifecycle_test.go (hunk 2 above)
Comment (ybrill) [reply model/lifecycle_test.go]: I reinstated the test

File: model/lifecycle_test.go (hunk 3 above)
Comment (ybrill) [reply model/lifecycle_test.go:1074]: I reinstated the test

File: model/task_lifecycle.go (hunk 4 above)
Comment (ybrill) [reply model/task_lifecycle.go]: done

**Comment by ybrill** [comment]
Thanks for pointing out the *.js issues. I think I missed a few places in the legacy front-end. 
They should be good now. The strategy I used was to populate [the uiTask slice](https://github.com/evergreen-ci/evergreen/blob/1f8eedfcec0126e298ed7ac1f201dc6b83bfb8be/service/models.go#L79) and change the js/html to use that instead of the task cache.

File: service/build.go [hunk 5]
```diff
@@ -24,12 +24,19 @@ import (
 
 // getUiTaskCache takes a build object and returns a slice of
 // uiTask objects suitable for front-end
-func getUiTaskCache(build *build.Build, tasks []task.Task) ([]uiTask, error) {
+func getUiTaskCache(b *build.Build) ([]uiTask, error) {
+	tasks, err := task.FindAll(task.ByBuildId(b.Id))
+	if err != nil {
+		return nil, errors.Wrap(err, "can't get tasks for build")
+	}
+	if tasks == nil {
```
Comment (ablack12) [inline service/build.go:32]: nit: can you check len(tasks) == 0 to cover both cases

File: service/build.go [hunk 6]
```diff
@@ -69,18 +76,19 @@ func (uis *UIServer) buildPage(w http.ResponseWriter, r *http.Request) {
...
+
+	timeTaken, makespan, err := projCtx.Build.GetTimeSpent()
 	if err != nil {
-		uis.LoggedError(w, r, http.StatusInternalServerError, err)
+		uis.LoggedError(w, r, http.StatusInternalServerError, errors.Wrap(err, "can't get time spent for build"))
 		return
 	}
-	buildAsUI.Tasks = uiTasks
-	buildAsUI.TimeTaken, buildAsUI.Makespan = task.GetTimeSpent(tasks)
+	buildAsUI.TimeTaken, buildAsUI.Makespan = timeTaken, makespan
```
Comment (ablack12) [inline service/build.go:91]: since err is defined at 79 already, you could just assign this directly at 86 and use =

File: service/build.go [hunk 7]
```diff
@@ -379,3 +413,23 @@ func (uis *UIServer) buildHistory(w http.ResponseWriter, r *http.Request) {
...
+func getBuildTaskMap(builds []build.Build) (map[string]map[string]task.Task, error) {
+	buildIds := make([]string, 0, len(builds))
+	for _, b := range builds {
+		buildIds = append(buildIds, b.Id)
+	}
+	tasksForBuilds, err := task.FindAll(task.ByBuildIds(buildIds).WithFields(task.BuildIdKey, task.DisplayNameKey, task.StatusKey, task.DetailsKey))
+	if err != nil {
+		return nil, errors.Wrap(err, "can't get tasks for builds")
+	}
+	buildTaskMap := make(map[string]map[string]task.Task)
```
Comment (ablack12) [inline service/build.go:426]: can you add some clarity comments about this data structure? Possibly a private type would be more readable, with two maps map[buildId][]string{taskIds} and map[taskId]task.Task{} but it is private so either does work

File: model/build/build.go [hunk 8]
```diff
@@ -247,6 +247,16 @@ func (b *Build) MarkFinished(status string, finishTime time.Time) error {
 	)
 }
 
+func (b *Build) GetTimeSpent() (time.Duration, time.Duration, error) {
+	tasks, err := task.FindAllFirstExecution(task.ByBuildId(b.Id).WithFields(task.TimeTakenKey, task.StartTimeKey, task.FinishTimeKey, task.DisplayOnlyKey, task.ExecutionKey))
```
Comment (ablack12) [inline model/build/build.go:251]: Is first execution what we want? I'd think we want the latest execution for each task.

**Review by ablack12 — CHANGES_REQUESTED** [review]

**Comment by ablack12** [comment]
Were you able to verify that the UI pages and restarting a task and etc still works in staging?

File: service/build.go (hunk 5 above)
Comment (ybrill) [reply service/build.go]: done

File: service/build.go (hunk 6 above)
Comment (ybrill) [reply service/build.go]: done

File: service/build.go (hunk 7 above)
Comment (ybrill) [reply service/build.go]: Splitting into two maps makes it much clearer.
I opted not to define a type.

File: model/build/build.go (hunk 8 above)
Comment (ybrill) [reply model/build/build.go:251]: The thinking of using the first execution is that when a task gets restarted the build's makespan shouldn't be the time from the start of the first task to run to the finish of the restarted task (potentially days).
This was the solution I landed on in an earlier PR (#3955).

**Comment by ybrill** [comment]
I did verify it in staging. The legacy UI is difficult to work with... 😣

File: model/build/build.go (hunk 8 above)
Comment (ablack12) [reply model/build/build.go:251]: Oh I see, that makes sense.

File: service/build.go [hunk 9]
```diff
@@ -414,22 +409,24 @@ func (uis *UIServer) buildHistory(w http.ResponseWriter, r *http.Request) {
...
 	for _, b := range builds {
 		buildIds = append(buildIds, b.Id)
 	}
 	tasksForBuilds, err := task.FindAll(task.ByBuildIds(buildIds).WithFields(task.BuildIdKey, task.DisplayNameKey, task.StatusKey, task.DetailsKey))
 	if err != nil {
-		return nil, errors.Wrap(err, "can't get tasks for builds")
+		return nil, nil, errors.Wrap(err, "can't get tasks for builds")
 	}
-	buildTaskMap := make(map[string]map[string]task.Task)
+	buildTaskIDsMap := make(map[string][]string)
```
Comment (ablack12) [inline service/build.go:424]: Actually follow up question: do we even need two maps? Could we do map[string][]task.Task{}? We aren't checking if the task already exists in the list at 427 anyway, I'm not sure we need the taskIds separately.

File: service/build.go [hunk 10]
```diff
@@ -294,18 +294,19 @@ func (uis *UIServer) modifyBuild(w http.ResponseWriter, r *http.Request) {
...
-	tasks, err := task.FindAllFirstExecution(task.ByBuildId(projCtx.Build.Id))
+	uiTasks, err := getUiTaskCache(projCtx.Build)
 	if err != nil {
 		uis.LoggedError(w, r, http.StatusInternalServerError, errors.Wrap(err, "can't get tasks for build"))
 		return
 	}
-	uiTasks, err := getUiTaskCache(projCtx.Build, tasks)
+	updatedBuild.Tasks = uiTasks
+
+	timeTaken, makespan, err := projCtx.Build.GetTimeSpent()
```
Comment (ablack12) [inline service/build.go:304]: Can you make the same change here?

Comment (ybrill) [reply service/build.go]: done

File: service/build.go (hunk 9 above)
Comment (ybrill) [reply service/build.go]: You're right. What I really intended to do was to look up the tasks from the task cache.
I've fixed it. 🤦 
Now it returns a single map from task id to task.

**Review by ablack12 — APPROVED** [review]

**Comment by ybrill** [comment]
@ablack12 I did a last minute check in staging and discovered that although the restart appeared to work the selected tasks weren't actually being restarted.
The reason was that the list of tasks we ask the server to restart was empty. The reason for this was that I was pulling the task ids to restart from the wrong place...
I put up 03352cb to fix it.

**Review by ablack12 — APPROVED** [review]

**Pull request merged**